### PR TITLE
std::format parsing callbacks and machinery

### PR DESCRIPTION
**Pull request by barcharcraz**
- adds `_Parse_format_string`
- adds `_Parse_replacement_field`
- adds `_Specs_setter` and `_Basic_format_specs`
- adds `_Specs_checker`, and `_Numeric_specs_checker`
- adds `_Format_handler` (note that `_On_format_specs` is incomplete)
- adds `vformat_to` definition
- adds `basic_format_args::get`

in total this is enough to get `vformat_to` instantiations to compile, although we can't run them without a few more `basic_format_arg` and `basic_format_args` related things.

Note that this `vformat_to` is the C++ 20 version of `vformat_to`, _NOT_ the proposed C++23 version (they have different ABIs)

Also note that this PR contains merge commits, they should go away if it's squashed, however.

### CODE REVIEW

File: stl/inc/format [hunk 1]
```diff
@@ -302,60 +402,77 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+    } _Writer_loop{_Handler};
+    while (_Begin != _End) {
+        const _CharT* _Pt = _Begin;
+        if (*_Begin != '{') {
+            _Pt = _STD find(_Begin + 1, _End, '{');
+            if (_Pt == _End) {
+                return _Writer_loop(_Begin, _End);
+            }
+        }
+        _Writer_loop(_Begin, _End);
```
Comment (cpplearner) [inline stl/inc/format:471]:
```suggestion
        _Writer_loop(_Begin, _Pt);

```

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/format [hunk 2]
```diff
@@ -443,13 +843,143 @@ public:
...
+    }
+
+    void _On_replacement_field(size_t _Id, const _CharT*) {
+        auto _Arg = _Ctx.arg(_Id);
+        _Ctx.advance_to(visit_format_arg(
+            _Default_arg_formatter<_OutputIt, _CharT>{_Ctx.out(), _Ctx._Get_args(), _Ctx.locale()}, _Arg));
+    }
+
+    const _CharT* _On_format_specs(size_t _Id, const _CharT* _Begin, const _CharT* _End) {
+        _Parse_context.advance_to(_Parse_context.begin() + (_Begin - &*_Parse_context.begin()));
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:951]: the character types are limited to char or wchar_t. I suppose address_of would be safer (I'll make the change), but I _think_ user defined op&s for char or wchar_t would be ambiguous in all cases right?

File: stl/inc/format [hunk 3]
```diff
@@ -389,26 +555,256 @@ public:
...
+// in addition handles dynamic width and precision. This is seperate from _Specs setter
+// because it needs to know about the current basic_format_parse_context and basic_format_context
+// in order to fetch the width from the arguments.
+template <class _ParseContext, class _Context>
+class _Specs_handler : public _Specs_setter<typename _Context::char_type> {
+public:
+    using _CharT = typename _Context::char_type;
+
+    constexpr _Specs_handler(_Basic_format_specs<_CharT>& _Specs, _ParseContext& _Parse_ctx, _Context& _Ctx)
+        : _Specs_setter<_CharT>(_Specs), _Parse_ctx(_Parse_ctx), _Ctx(_Ctx) {}
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:691]: I conventionally almost always shadow members like when when I'm just going to be using them in a member initializer list.

I guess it could cause problems if the constructor ever becomes more than just a member initializer list

File: stl/inc/format [hunk 4]
```diff
@@ -389,26 +555,256 @@ public:
...
+    constexpr basic_format_arg<_Context> _Get_arg(size_t _Arg_id) {
+        _Parse_ctx.check_arg_id(_Arg_id);
+        return _STD _Get_arg(_Ctx, _Arg_id);
+    }
 };
+class _Numeric_specs_checker {
+    _Type _Arg_type = _Type::_None;
+
+public:
+    constexpr _Numeric_specs_checker(_Type _Arg_type) : _Arg_type(_Arg_type) {}
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:720]: sure, why not!

File: stl/inc/format (hunk 2 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/format:951]: It's impossible to provide a non-member overload of `operator&` for `char`/`wchar_t`. I recommend using `&` for these known character types - `addressof` has debug codegen costs. (We conventionally avoid `addressof` for known builtin types.)

File: stl/inc/format [hunk 5]
```diff
@@ -38,36 +40,122 @@ enum class _Align { _None, _Left, _Right, _Center };
...
+    // _Next_arg_id == 0 means unknown
+    // _Next_arg_id > 0 means automatic
+    // _Next_arg_id == -1 means manual
+    constexpr size_t next_arg_id() {
+        if (_Next_arg_id >= 0) {
+            return static_cast<size_t>(_Next_arg_id++);
+        }
+        throw format_error("Can not switch from manual to automatic indexing");
+    }
+    constexpr void check_arg_id(size_t _Id) {
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/format:113]: I believe this part should be dropped in favor of #1597

File: stl/inc/format [hunk 6]
```diff
@@ -389,26 +555,256 @@ public:
...
+    }
+};
+
+template <class _Ty, class _CharT = char>
+struct formatter;
+
 
 // TODO: test coverage
 template <class _Context, class... _Args>
 class _Format_arg_store {
```
[the remark this answers is not in the document]
Comment (cpplearner) [reply stl/inc/format:783]: The standard says

      template<class Context, class... Args> struct format-arg-store;      // exposition only

Does that permit an evil program to match its template parameter list with that of a template template argument?

File: stl/inc/format (hunk 5 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:113]: yeah, although this appears in green here mostly because the class was moved upwards in the file

File: stl/inc/format (hunk 2 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:951]: IF character types must be either `char` or `wchar_t` then we're good here.

File: stl/inc/format (hunk 6 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:783]: Yes, but I'm willing to fix the Standard if something in the wording prevents us making our implementation as efficient as otherwise possible for no good reason. (This comment is likely moot anyway given #1625.)

File: tests/std/tests/P0645R10_text_formatting_formatting/env.lst [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_formatting/env.lst:1]: yeah, format is also growing more independent test cases than I'd really like.

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/tests/P0645R10_text_formatting_formatting/env.lst:1]: > yeah, format is also growing more independent test cases than I'd really like.

If you'd prefer a single monolithic test, then say so. My sole concern here is that the tests actually run ;)

File: stl/inc/format [hunk 8]
```diff
@@ -188,17 +275,30 @@ struct _Precision_adapter {
...
-    constexpr void _On_manual_id(int _Id) {
+    constexpr void _On_manual_id(size_t _Id) {
         _Callbacks._On_dynamic_precision(_Id);
     }
 };
 
+template <class _CharT>
+struct _Id_adapter {
+    basic_format_parse_context<_CharT>& _Parse_context;
+    size_t _Arg_id = 0;
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:286]: sure why not. But invalid arg_ids should fail to parse _anyways_

if we do that we should add an assert after calling the parsing function as well.

File: stl/inc/format [hunk 9]
```diff
@@ -302,60 +402,77 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+constexpr void _Parse_format_string(basic_string_view<_CharT> _Format_str, _HandlerT&& _Handler) {
+    auto _Begin = _Format_str.data();
+    auto _End   = _Begin + _Format_str.size();
+    struct _Writer {
+        constexpr void operator()(const _CharT* _Begin, const _CharT* _End) {
+            if (_Begin == _End) {
+                return;
+            }
+            for (;;) {
+                const _CharT* _Pt = _STD find(_Begin, _End, '}');
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:449]: libfmt does just call memchr, but it also special cases for small numbers of characters (to avoid the function call) and has a specialization for wchar_t format strings that's a simple loop.

I figure it's probably easier / better to just call find (or find_unchecked)

File: stl/inc/format [hunk 10]
```diff
@@ -364,23 +481,72 @@ public:
...
 private:
-    using _Char_type = typename _Context::char_type;
+    template <class _Visitor, class _Ctx>
+    friend auto visit_format_arg(_Visitor&& _Vis, basic_format_arg<_Ctx> _Arg) -> decltype(_Vis(0));
+
+    using _CharT = typename _Context::char_type;
 
     using _Format_arg_value = variant<monostate, int, unsigned, long long, unsigned long long, bool,
         typename _Context::char_type, float, double, long double, const void*, const typename _Context::char_type,
         basic_string_view<typename _Context::char_type>, handle>;
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:491]: not going to apply suggestion since #1625 supersedes it

File: stl/inc/format [hunk 11]
```diff
@@ -389,26 +555,256 @@ public:
     explicit operator bool() const noexcept {
         return !_STD holds_alternative<monostate>(_Value);
     }
+    constexpr _Type _Get_type() {
+        return _STD visit(_Format_arg_type_to_enum<_CharT>{}, _Value);
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:559]: I thought about that, also this will be done with a custom variant type in #1625 anyways

File: stl/inc/format [hunk 12]
```diff
@@ -389,26 +555,256 @@ public:
...
+    constexpr void _On_width(unsigned int _Width) {
+        _Specs._Width = _Width;
+    }
+
+    constexpr void _On_precision(unsigned int _Precision) {
+        _Specs._Precision = _Precision;
+    }
+
+    constexpr void _On_type(_CharT _Type) {
+        _Specs._Type = static_cast<_CharT>(_Type);
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:614]: libfmt actually casts to char here (and the type of _Type is char), however we may need to keep the type wide for execution charsets that don't map wide and narrow latin characters to the same codepoint :(.

File: stl/inc/format [hunk 13]
```diff
@@ -398,26 +570,256 @@ public:
...
+                }
+            }
+            return static_cast<unsigned long long>(_Value);
+        }
+        throw format_error("Precision is not an integer.");
+    }
+};
+
+// Parses standard format specs into a _Basic_format_specs using _Specs_setter, and,
+// in addition handles dynamic width and precision. This is seperate from _Specs setter
```
Comment (StephanTLavavej) [inline stl/inc/format:697]:
```suggestion
// in addition handles dynamic width and precision. This is separate from _Specs setter
```

File: stl/inc/format [hunk 14]
```diff
@@ -398,26 +570,256 @@ public:
...
     basic_format_args() noexcept;
     template <class... _Args>
     basic_format_args(const _Format_arg_store<_Context, _Args...>& store) noexcept;
 
-    basic_format_arg<_Context> get(size_t _Index) const noexcept;
+    basic_format_arg<_Context> get(size_t _Index) const noexcept {
+        if (_Index < _Size) {
+            return _Data[_Index];
+        } else {
+            return basic_format_arg<_Context>();
```
Comment (StephanTLavavej) [inline stl/inc/format:820]: In new code we try to use empty braces to construct temporaries (as this avoids looking like a function call).

File: stl/inc/format [hunk 15]
```diff
@@ -443,13 +843,143 @@ public:
...
+
+template <class _Visitor, class _Context>
+auto visit_format_arg(_Visitor&& _Vis, basic_format_arg<_Context> _Arg) -> decltype(_Vis(0)) {
+    return _STD visit(_STD forward<_Visitor>(_Vis), _Arg._Value);
+}
+
+template <class _CharT, class _OutputIt>
+_OutputIt _Write(_OutputIt _Out, monostate) {
+    _STL_INTERNAL_CHECK(false);
+    return _Out;
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/format:875]: Could this perhaps be `=delete`d?

File: stl/inc/format [hunk 16]
```diff
@@ -452,13 +858,143 @@ public:
...
+    using _Char_type = typename _Context::char_type;
+
+    basic_format_parse_context<_Char_type>& _Parse_ctx;
+    _Context& _Ctx;
+
+public:
+    explicit constexpr _Custom_formatter_dispatcher(basic_format_parse_context<_Char_type>& _Parse_ctx, _Context& _Ctx)
+        : _Parse_ctx(_Parse_ctx), _Ctx(_Ctx) {}
+
+    void operator()(typename basic_format_arg<_Context>::handle h) const {
```
Comment (StephanTLavavej) [inline stl/inc/format:909]: `h` is not `_Ugly`.

File: stl/inc/format [hunk 17]
```diff
@@ -302,69 +418,76 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
-    // _Next_arg_id == -1 means manual
-    constexpr size_t next_arg_id() {
-        if (_Next_arg_id == -1) {
-            throw format_error("Can not switch from manual to automatic indexing");
+    if (*_Begin == '}') {
+        // string was "{}", and we have a replacement field
+        _Handler._On_replacement_field(_Handler._Parse_context.next_arg_id(), _Begin);
+    } else if (*_Begin == '{') {
+        // string was "{{", so we have a literal "{" to print
+        _Handler._On_text(_Begin, _Begin + 1);
```
Comment (miscco) [inline stl/inc/format:433]: If I read that correct `_Begin` is still pointing to the first `{` So if we then return `_Begin + 1` we would try to parse that as a format string.

TL;DR: Should we increment `_Begin` here as this is essentially a character of width 2?

File: stl/inc/format [hunk 18]
```diff
@@ -302,69 +418,76 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+    } else {
+        _Id_adapter<_CharT> _Adapter{_Handler._Parse_context};
+        _Begin     = _Parse_arg_id(_Begin, _End, _Adapter);
+        _CharT _Ch = _CharT{};
+        if (_Begin != _End) {
+            _Ch = *_Begin;
+        }
```
Comment (miscco) [inline stl/inc/format:440]: I think that if `_Begin == _End` we will always end up in the throw error case.

At least for me It would be clearer if we would write it like that
```suggestion
        _CharT _Ch = _CharT{};
        if (_Begin != _End) {
            _Ch = *_Begin;
        }
```suggestion
        if (_Begin == _End) {
            throw format_error("Missing '}' in format string.");
        }
        
        
       const  _CharT _Ch = *_Begin;
```
```

File: stl/inc/format [hunk 19]
```diff
@@ -302,69 +418,76 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+}
 
-    constexpr void check_arg_id(const size_t _Id) {
-        if (_STD is_constant_evaluated()) {
-            if (_Id >= _Num_args) {
-                _You_see_this_error_because_arg_id_is_out_of_range();
+template <class _CharT, _Parse_replacement_field_callbacks<_CharT> _HandlerT>
+constexpr void _Parse_format_string(basic_string_view<_CharT> _Format_str, _HandlerT&& _Handler) {
+    auto _Begin = _Format_str.data();
+    auto _End   = _Begin + _Format_str.size();
```
Comment (miscco) [inline stl/inc/format:458]: I think it would be clearer if we would do
```suggestion
    auto _Begin = _Format_str._Unchecked_begin();
    auto _End   = _Format_str._Unchecked_end();;
```

File: stl/inc/format [hunk 20]
```diff
@@ -302,69 +418,76 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+    struct _Writer {
+        constexpr void operator()(const _CharT* _Begin, const _CharT* _End) {
+            if (_Begin == _End) {
+                return;
+            }
+            for (;;) {
+                const _CharT* _Pt = _STD _Find_unchecked(_Begin, _End, '}');
+                if (_Pt == _End) {
+                    return _Handler._On_text(_Begin, _End);
```
Comment (miscco) [inline stl/inc/format:467]: Is that really correct, we did not really find the `}` in this case didnt we?

File: stl/inc/format [hunk 21]
```diff
@@ -302,69 +418,76 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
 
-    constexpr void check_arg_id(const size_t _Id) {
-        if (_STD is_constant_evaluated()) {
-            if (_Id >= _Num_args) {
-                _You_see_this_error_because_arg_id_is_out_of_range();
+template <class _CharT, _Parse_replacement_field_callbacks<_CharT> _HandlerT>
+constexpr void _Parse_format_string(basic_string_view<_CharT> _Format_str, _HandlerT&& _Handler) {
+    auto _Begin = _Format_str.data();
+    auto _End   = _Begin + _Format_str.size();
+    struct _Writer {
```
Comment (miscco) [inline stl/inc/format:459]: Is there a reason that we define this inside here?

File: stl/inc/format (hunk 17 above)
Comment (cpplearner) [reply stl/inc/format:433]: No, because this function returns `_Begin + 1`.

Comment (cpplearner) [reply stl/inc/format:433]: `_Begin` is pointing to the second `{`: it points to the first `{` before entering this function, and the first thing this function does is `++_Begin`.

File: stl/inc/format (hunk 20 above)
Comment (cpplearner) [reply stl/inc/format:467]: It's normal that `}` is not found, e.g. `format("xxx")`.

Below handles two consecutive `}`, e.g. in `format("xxx }} xxx")`, in which case a single `}` is written.

File: stl/inc/format [hunk 22]
```diff
@@ -38,36 +40,122 @@ enum class _Align { _None, _Left, _Right, _Center };
...
+    _Float32,
+    _Float64,
+    _Last_numeric_type = _Float64,
+    _CString,
+    _String,
+    _Pointer,
+    _Custom
+};
+constexpr bool _Is_integral_fmt_type(_Type _Ty) {
+    return _Ty > _Type::_None && _Ty <= _Type::_Last_integer_type;
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:61]: resolved by moveing to _Basic_format_arg_type and just using the explicit types without marker enumerators.

File: stl/inc/format [hunk 23]
```diff
@@ -38,36 +40,122 @@ enum class _Align { _None, _Left, _Right, _Center };
...
     { _At._On_auto_id() } -> same_as<void>;
-    { _At._On_manual_id(_STD declval<int>()) } -> same_as<void>;
+    { _At._On_manual_id(size_t{}) } -> same_as<void>;
+};
+
+template<class _Ty, class _CharT>
+concept _Parse_replacement_field_callbacks = requires(_Ty _At, const _CharT* _Begin, const _CharT* _End) {
+    { _At._Parse_context };
+    { _At._On_text(_Begin, _End) } -> same_as<void>;
+    { _At._On_replacement_field(size_t{}, _STD declval<const _CharT*>()) } -> same_as<void>;
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:158]: yeah sure why not.

File: stl/inc/format [hunk 24]
```diff
@@ -389,26 +555,256 @@ public:
...
+};
+
+template <typename _CharT>
+struct _Basic_format_specs {
+    unsigned int _Width, _Precision;
+    char _Type;
+    _Align _Alignment;
+    _Sign _Sgn;
+    bool _Alt;
+    _CharT _Fill[4];
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:585]: we're pretty sure

That's how libfmt interprets it.

File: stl/inc/format [hunk 25]
```diff
@@ -389,26 +555,256 @@ public:
...
+class _Specs_setter {
+public:
+    explicit constexpr _Specs_setter(_Basic_format_specs<_CharT> _Specs) : _Specs(_Specs) {}
+
+    constexpr void _On_align(_Align _Aln) {
+        _Specs._Alignment = _Aln;
+    }
+
+    constexpr void _On_fill(basic_string_view<_CharT> _Sv) {
+        if (_Sv.size() > 4) {
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:600]: later, may need to change it to a class anyways

File: stl/inc/format [hunk 26]
```diff
@@ -389,26 +555,256 @@ public:
...
+
+protected:
+    _Basic_format_specs<_CharT> _Specs;
+};
+
+// Actually fetch the value of an argument associated with a dynamic
+// width or precision specifier. This will be called with either
+// _Width_checker or _Precision_checker as "_Handler".
+template <class _Handler, class _FormatArg>
+constexpr unsigned int _Get_dynamic_specs(_FormatArg _Arg) {
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:640]: it could, but that would only save 9 bytes, since basic_format_arg is a 1 byte enum + at most a basic_string_view

File: stl/inc/format [hunk 27]
```diff
@@ -389,26 +555,256 @@ public:
...
+        : _Handler(_Handler_inst), _Numeric_checker(_Arg_type) {}
+
+    // _On_align has no checking, since we don't implement numeric alignments.
+    constexpr void _On_sign(_Sign _Sgn) {
+        _Numeric_checker._Check_sign();
+        _Handler::_On_sign(_Sgn);
+    }
+
+    constexpr void _On_hash() {
+        _Numeric_checker._Require_numeric_argument();
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:777]: no, it doesn't. That has to be checked later since you can have a hash with a bool or char if you've passed a numeric "type" later on. I'll add a comment

File: stl/inc/format (hunk 15 above)
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:872]: it does not seem to cause warnings, and we can't delete it because visit_format_arg switches based on the real, actual, runtime type of the type enumerator, resulting in the instantiation of all the alternative paths

File: stl/inc/format [hunk 28]
```diff
@@ -443,13 +843,143 @@ public:
...
+
+    explicit _Format_handler(
+        _OutputIt _Out, basic_string_view<_CharT> _Str, basic_format_args<_Context> _Format_args, const locale& _Loc)
+        : _Parse_context(_Str), _Ctx(_Out, _Format_args, _Loc) {}
+
+    void _On_text(const _CharT* _Begin, const _CharT* _End) {
+        auto _Size = _End - _Begin;
+        auto _Out  = _Ctx.out();
+        _STD copy_n(_Begin, _Size, _Out);
+        _Ctx.advance_to(_Out);
```
[the remark this answers is not in the document]
Comment (barcharcraz) [reply stl/inc/format:953]: yep, it's a string_view call to _Write in libfmt

File: stl/inc/format (hunk 19 above)
Comment (barcharcraz) [reply stl/inc/format:576]: it's minor so I'll keep as is, just to preserve similarity with libfmt (and because I wanna get this pr merged)

File: stl/inc/format (hunk 15 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:1115]: AFAICS users can write `std::visit_format_arg(my_visitor, basic_format_arg<SomeContextType>{})` which results in this function being called, so we should eventually have test coverage for it instead and the `_STL_INTERNAL_CHECK(false)` will have to go away in any case. I don't much care whether we remove it now vs. later, so I'll retract this comment.

File: stl/inc/format [hunk 29]
```diff
@@ -38,36 +40,248 @@ enum class _Align { _None, _Left, _Right, _Center };
...
+    _NODISCARD constexpr const_iterator begin() const noexcept {
+        return _Format_string.begin();
+    }
+    _NODISCARD constexpr const_iterator end() const noexcept {
+        return _Format_string.end();
+    }
+    constexpr void advance_to(const const_iterator _It) {
+        _Adl_verify_range(_It, _Format_string.end());
+        const auto _Diff = static_cast<size_t>(_It._Unwrapped() - _Format_string._Unchecked_begin());
+        _Format_string.remove_prefix(_Diff);
```
Comment (CaseyCarter) [inline stl/inc/format:127]: I'm concerned that `_Diff` could be negative here and `remove_prefix` will do the wrong thing (it takes a `size_t`). We could we comment this for now and worry about it later.

File: stl/inc/format (hunk 24 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/format:624]: This should probably be `MB_LEN_MAX` to properly support all narrow locales, but we can worry about that later.

File: stl/inc/format [hunk 30]
```diff
@@ -302,202 +536,308 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+    auto _Begin = _Format_str.data();
+    auto _End   = _Begin + _Format_str.size();
+    while (_Begin != _End) {
+        const _CharT* _OpeningCurl = _Begin;
+        if (*_Begin != '{') {
+            // we didn't start at an opening curl, find the next one
+            _OpeningCurl = _STD find(_Begin + 1, _End, '{');
+            _STL_INTERNAL_CHECK(_Begin != _OpeningCurl);
+            for (;;) {
+                const _CharT* _ClosingCurl = _STD _Find_unchecked(_Begin, _End, '}');
```
Comment (CaseyCarter) [inline stl/inc/format:584]: We don't (yet?) `_STD`-qualify ugly function calls. Feel free to merge as-is and fix later.

File: stl/inc/format [hunk 31]
```diff
@@ -302,202 +536,308 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+template <class _CharT, _Parse_replacement_field_callbacks<_CharT> _HandlerT>
+constexpr void _Parse_format_string(basic_string_view<_CharT> _Format_str, _HandlerT&& _Handler) {
+    auto _Begin = _Format_str.data();
+    auto _End   = _Begin + _Format_str.size();
+    while (_Begin != _End) {
+        const _CharT* _OpeningCurl = _Begin;
+        if (*_Begin != '{') {
+            // we didn't start at an opening curl, find the next one
+            _OpeningCurl = _STD find(_Begin + 1, _End, '{');
+            _STL_INTERNAL_CHECK(_Begin != _OpeningCurl);
```
Comment (CaseyCarter) [inline stl/inc/format:582]: Odd place to check for bugs in `find`.

File: stl/inc/format [hunk 32]
```diff
@@ -302,202 +536,308 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+            for (;;) {
+                const _CharT* _ClosingCurl = _STD _Find_unchecked(_Begin, _End, '}');
+
+                // In this case we didn't find either a closing curl or opening curl.
+                // Write the whole thing out.
+                if (_ClosingCurl == _OpeningCurl) {
+                    return _Handler._On_text(_Begin, _OpeningCurl);
+                }
+                // We know _ClosingCurl isn't past the end because
+                // the above condition was not met.
```
Comment (CaseyCarter) [inline stl/inc/format:592]: Either I don't understanding or this statement is just wrong. `'x{{"` is a valid format-string for which `_OpeningCurl` is not equal to `_ClosingCurl` yet `_ClosingCurl` is past-the-end here. Does this function have some undocumented precondition?

File: stl/inc/format [hunk 33]
```diff
@@ -302,202 +536,308 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+                ++_ClosingCurl;
+                if (_ClosingCurl == _OpeningCurl || *_ClosingCurl != '}') {
+                    throw format_error("Unmatched '}' in format string.");
+                }
+                // We found two closing curls, so output online one of them
```
Comment (CaseyCarter) [inline stl/inc/format:597]:
```suggestion
                // We found two closing curls, so output only one of them
```

File: stl/inc/format [hunk 34]
```diff
@@ -302,202 +536,308 @@ constexpr const _CharT* _Parse_format_specs(const _CharT* _Begin, const _CharT*
...
+        return _STD _Get_arg(_Ctx, _Arg_id);
+    }
+
+    // Fetch the value of an argument associated with a dynamic
+    // width or precision specifier. This will be called with either
+    // _Width_checker or _Precision_checker as "_Handler".
+    template <class _Handler, class _FormatArg>
+    static constexpr unsigned int _Get_dynamic_specs(_FormatArg _Arg) {
+        unsigned long long _Val = _STD visit_format_arg(_Handler(), _Arg);
+        if (_Val > (_STD numeric_limits<unsigned int>::max)()) {
```
Comment (CaseyCarter) [inline stl/inc/format:765]: No need to `_STD`-qualify the name of the class template `numeric_limits`.

**Review by CaseyCarter — APPROVED** [review]
Feel free to merge as-is and make changes in a later PR.
